File: maxHeap.cpp
#include "iostream"
using namespace std;

// Same as min heap just the less than in min heap is swithed to greater than in max heap

struct heap{
	int *ar;
	int capacity,cur;
};

heap* heapInit(){
	heap *ptr = new heap;
	ptr->cur = 0;
	ptr->capacity = 50;
	ptr->ar = new int[50];
	return ptr;
}

void swap(int &a,int &b){
	int t=a;a=b;b=t;
}

void heapifyBT(heap *root,int index){
	int parent = (index - 1) / 2;
	if(root->ar[parent] <  root->ar[index]){
		swap(root->ar[parent],root->ar[index]);
		heapifyBT(root,parent);
	}
}

void heapifyTB(heap *root,int parent){
	int left = parent * 2 + 1;
	int right = parent * 2 + 2;
	int max = parent;

	if(left >= root->cur || left < 0)
		left = -1;

	if(right >= root->cur || right < 0)
		right = -1;
	
	if(left != -1 && root->ar[left] > root->ar[parent])// point of difference from min heap
		max = left;

	if(right != -1 && root->ar[right] > root->ar[max])// point of difference from min heap
		max = right;

	if(max != parent){
		swap(root->ar[parent],root->ar[max]);
		heapifyTB(root,max);
	}
}

void insert(heap *root,int key){
	if(root->cur < root->capacity){// point of difference from min heap
		root->ar[root->cur] = key;
		heapifyBT(root,root->cur);
		root->cur += 1;
	}
}

void show(heap *root){
	for(int i=0;i<root->cur;++i)
		cout<<root->ar[i]<<" ";
	cout<<"\n";
}

void popMax(heap *root){
	if(root->cur){
		root->ar[0] = root->ar[root->cur - 1];
		root->cur -= 1;
		heapifyTB(root,0);
	}
	else
		cout<<"heap is empty\n";
}

int main(){
	int n,key;
	cout<<"total elements: ";
	cin>>n;

	heap *root = heapInit();
	for(int i=0;i<n;++i){
		cin>>key;
		insert(root,key);
	}

	show(root);
	popMax(root);
	show(root);
}

File: DSA_in_JAVA/maxHeap.java
import java.util.*;

class heaps
{
  int cur,cap;
  int ar[];

  heaps()
  {
    cur = 0;
    cap = 50;
    ar = new int[cap];
  }
}

public class maxHeap {
  static void show(heaps hp)
  {
    if(hp.cur > 0)
    {
      for(int i = 0;i<hp.cur;i++)
        System.out.print(hp.ar[i]+" ");
      System.out.println();
    }
    else
      System.out.println("Empty Heap");
  }

  static void heapifyTB(heaps hp,int ind)
  {
    int l = ind * 2 + 1;
    int r = ind * 2 + 2;
    int m = ind;

    if(l < hp.cur && hp.ar[l] > hp.ar[m])
      m = l;
    if(r < hp.cur && hp.ar[r] > hp.ar[m])
      m = r;
    
    if(m != ind)
    {
      int t = hp.ar[m];
      hp.ar[m] = hp.ar[ind];
      hp.ar[ind] = t;
      heapifyTB(hp,m);
    }
  }

  static void heapifyBT(heaps hp,int ind)
  {
    int p = (ind - 1) / 2;

    if(hp.ar[ind] > hp.ar[p])
    {
      int t = hp.ar[p];
      hp.ar[p] = hp.ar[ind];
      hp.ar[ind] = t;
      heapifyBT(hp,p);
    }
  }

  static void insert(heaps hp,int v)
  {
    if(hp.cur < hp.cap)
    {
      hp.ar[hp.cur] = v;
      heapifyBT(hp,hp.cur);
      hp.cur++;
    }
  }

  static void pop(heaps hp)
  {
    if(hp.cur > 0)
    {
      hp.ar[0] = hp.ar[--hp.cur];
      heapifyTB(hp,0);
    }
  }

  public static void main(String[] args) {
    Scanner scan = new Scanner(System.in);
    int n = scan.nextInt();

    heaps hp = new heaps();

    for(int i=0;i<n;i++)
    {
      int v = scan.nextInt();
      insert(hp,v);
    }
    show(hp);
    pop(hp);
    show(hp);
  }  
}
